core/services/workflows/syncer/engine_registry.go

```diff
@@ -12,64 +13,112 @@ import (
 
 var errNotFound = errors.New("engine not found")
 
+type EngineRegistryKey struct {
+	Owner []byte
+	Name  string
+}
+
+// KeyFor generates a key that will be used to identify the engine in the engine registry.
+// This is used instead of a Workflow ID, because the WID will change
+// if the workflow code is modified.
+func (k EngineRegistryKey) keyFor() string {
+	return hex.EncodeToString(k.Owner) + "-" + name
+}
+
+type ServiceWithMetadata struct {
+	workflowID    [32]byte
+	workflowName  string
+	workflowOwner []byte
+	services.Service
+}
+
 type EngineRegistry struct {
-	engines map[string]services.Service
+	engines map[string]ServiceWithMetadata
 	mu      sync.RWMutex
 }
 
 func NewEngineRegistry() *EngineRegistry {
 	return &EngineRegistry{
-		engines: make(map[string]services.Service),
+		engines: make(map[string]ServiceWithMetadata),
 	}
 }
 
 // Add adds an engine to the registry.
-func (r *EngineRegistry) Add(id string, engine services.Service) error {
+func (r *EngineRegistry) Add(key EngineRegistryKey, engine services.Service, workflowID [32]byte) error {
 	r.mu.Lock()
 	defer r.mu.Unlock()
-	if _, found := r.engines[id]; found {
+
+	k := key.keyFor()
+	if _, found := r.engines[k]; found {
 		return errors.New("attempting to register duplicate engine")
 	}
-	r.engines[id] = engine
+	r.engines[k] = ServiceWithMetadata{
+		workflowID:    workflowID,
+		workflowName:  key.Name,
+		workflowOwner: key.Owner,
+		Service:       engine,
+	}
 	return nil
 }
 
 // Get retrieves an engine from the registry.
-func (r *EngineRegistry) Get(id string) (services.Service, error) {
+func (r *EngineRegistry) Get(key EngineRegistryKey) (ServiceWithMetadata, error) {
 	r.mu.RLock()
 	defer r.mu.RUnlock()
-	engine, found := r.engines[id]
+
+	k := key.keyFor()
```

Comment:
Nits throughout this file: a lot of these could be inlined with their use if they're just used in one place

Also think of line breaks as paragraphs; they cluster or don't cluster together units of behaviour. When you think about it that way, what value is there in having this isolated from its use two lines down?

Reply:
Yes, fair nits. Cleaned this up.